### ISSUE TITLE
Scope: Mixed per-axis boundary conditions in 3D

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/spectraldiffx#43]
## Summary

Scope and implement support for **mixed per-axis boundary conditions in 3D** — e.g., periodic in x and y, Neumann in z. This extends the 2D mixed-BC solver from #39 to three dimensions.

## Background

Many 3D physical problems have different BCs per axis:
- **Atmospheric boundary layer:** periodic x/y, Neumann z (ground + top)
- **Channel flow:** periodic x/z, Dirichlet y (walls)
- **Ocean models:** periodic x/y, mixed Dirichlet/Neumann z (surface/bottom)
- **Duct flow:** Dirichlet y/z (walls), periodic or inflow/outflow x

PoisFFT implements this for 3D (Section 2.3, 3) using a **pencil decomposition** approach:
1. Transform the periodic axes first (2D FFT in x-y)
2. For each (kx, ky) wavenumber pair, solve a 1D Helmholtz problem in z using the appropriate DST/DCT
3. Inverse transform the periodic axes

The effective 1D Helmholtz parameter for each (kx, ky) mode is:
```
λ_eff(kx, ky) = λ - eigx[kx] - eigy[ky]
```

PoisFFT's CLMM application (Section 5) uses exactly this: periodic x, periodic or Neumann y, Neumann z — for pressure correction in atmospheric LES.

## Needs Scoping

1. **Pencil decomposition strategy for 3D:** When all three axes have different BC types, we need nested pencil solves. Determine the optimal ordering (which axes to batch-transform first).

2. **Complex intermediate handling:** When mixing FFT (complex) with DST/DCT (real), the intermediate array after the FFT is complex. The 1D DST/DCT solve must handle complex inputs (apply separately to real and imaginary parts, or restructure the solve).

3. **API design for 3D:**
   ```python
   solve_helmholtz_3d(
       rhs, dx, dy, dz,
       bc_x="periodic", bc_y="periodic", bc_z="neumann_stag",
       lambda_=0.0,
   )
   ```

4. **Performance considerations:** The pencil approach requires sequential 1D transforms rather than a single 3D transform. For the common case of 2 periodic + 1 non-periodic axis, we can use a 2D FFT + batched 1D DST/DCT, which is efficient. For 3 different BC types, we need 3 sequential 1D passes.

5. **Module class design:**
   ```python
   class MixedBCHelmholtzSolver3D(eqx.Module):
       dx: float; dy: float; dz: float
       bc_x: str; bc_y: str; bc_z: str
       alpha: float = 0.0
   ```

6. **Test cases:** Reproduce PoisFFT's 3D convergence results (Section 4.1, Fig. 5 bottom row) for periodic, Dirichlet, and Neumann BCs. Add mixed-BC tests with known analytic solutions.

7. **vmap compatibility:** The batched 1D solves should work with `jax.vmap` for efficiency. Verify this works with the 1D functional solvers from #38.

## Depends On

- #38 (1D eigenvalue functions and solvers)
- #39 (2D mixed-BC solver — the 3D version extends the same approach)

## References

- PoisFFT paper Sections 2.3, 3, 5 (3D mixed BCs, pencil decomposition, CLMM application)
- SailFFish paper Section 2.6 (semi-unbounded solvers — planned approach for mixed 3D)
- FLUPS library: Caprace et al. (2021), "A Fourier-based library of unbounded Poisson solvers" — reference implementation for mixed 3D BCs
- Full plan: `spectraldiffx/.plans/staggered-grid-solvers.md` (Open Question 4)
